Widgets/Widgets/PictureOfTheDayWidget.swift

```diff
@@ -0,0 +1,248 @@
+
+	// MARK: Public
+
+	func fetchLatestAvailablePictureEntry(usingImageCache: Bool = false, _ completion: @escaping (PictureOfTheDayEntry) -> Void) {
+		// We could Result type the completion, but it's not probably worth the complexity for the widget's use cases
+		guard let contentGroup = dataStore.viewContext.newestGroup(of: .pictureOfTheDay), let imageContent = contentGroup.contentPreview as? WMFFeedImage else {
+			completion(sampleEntry)
+			return
+		}
+
+		let sampleEntry = self.sampleEntry
+		let contentDate = contentGroup.date
+		let contentURL = contentGroup.url
+		let imageThumbnailURL = imageContent.imageThumbURL
+		let imageDescription = imageContent.imageDescription
+
+		guard !usingImageCache else {
+			if let cachedImage = ImageCacheController.shared?.memoryCachedImage(withURL: imageThumbnailURL) {
+				let entry = PictureOfTheDayEntry(date: Date(), contentDate: contentDate, contentURL: contentURL, imageURL: imageThumbnailURL, image: cachedImage.staticImage, imageDescription: imageDescription)
+				completion(entry)
+			} else {
+				completion(sampleEntry)
+			}
+			return
+		}
+
+		ImageCacheController.shared?.fetchImage(withURL: imageThumbnailURL, failure: { _ in
+			completion(sampleEntry)
+		}, success: { fetchedImage in
+			let entry = PictureOfTheDayEntry(date: Date(), contentDate: contentDate, contentURL: contentURL, imageURL: imageThumbnailURL, image: fetchedImage.image.staticImage, imageDescription: imageDescription)
+			completion(entry)
+		})
+	}
+
+	// MARK: Private
+
+	private func fetchImageLicense(forCanonicalPageTitle imageFile: String, _ completion: @escaping (MWKImageInfo?) -> Void) {
+		imageInfoFetcher.fetchImageInfo(forCommonsFiles: [imageFile], failure: { error in
+			completion(nil)
+		}, success: { infoArray in
+			guard let infoArray = infoArray as? [MWKImageInfo] else {
+				completion(nil)
+				return
+			}
+
+			if let _ = infoArray.first {
+				// TODO: info.license
+			} else {
+				completion(nil)
+			}
+		})
+	}
+
+}
+
+// MARK: - Model
+
+struct PictureOfTheDayEntry: TimelineEntry {
+	let date: Date // for Timeline Entry
+	var contentDate: Date? = nil
+	var contentURL: URL? = nil
+	var imageURL: URL? = nil
+	let image: UIImage?
+	var imageDescription: String? = nil
+	var license: MWKLicense? = nil
+}
+
+// MARK: - TimelineProvider
+
+struct PictureOfTheDayProvider: TimelineProvider {
+
+	// MARK: Nested Types
+	
+	public typealias Entry = PictureOfTheDayEntry
+
+	// MARK: Properties
+	
+	private let dataStore = PictureOfTheDayData.shared
+
+	// MARK: TimelineProvider
+
+	func placeholder(in: Context) -> PictureOfTheDayEntry {
+		return dataStore.placeholderEntry
+	}
+    
+	func getTimeline(in context: Context, completion: @escaping (Timeline<PictureOfTheDayEntry>) -> Void) {
+		dataStore.fetchLatestAvailablePictureEntry { entry in
+			let currentDate = Date()
+			let timeline = Timeline(entries: [entry], policy: .after(currentDate.dateAtMidnight() ?? currentDate))
```

Comment:
I'm not dealing with it yet, but been thinking about this for the widget I'm working on, and need to dig into this to investigate more.

It's not always just midnight, right? I'm not sure when these refresh on the server. But even if it's midnight in some timezone, no guarantees the phone is in that timezone. This is definitely more pressing for "On This Day", but I'm going to need to figure out when we can get refreshed data so we're not surfacing day-old "On This Day" data. (I'm just providing comments as I think of them at this point - not a full review yet.)

Reply:
Thanks for peeking at all this. 

The way I understand the [Picture of the day on Commons](https://commons.wikimedia.org/wiki/Commons:Picture_of_the_day) is that they are often determined much in advance. So today, I could potentially request [the picture for a week from now](https://en.wikipedia.org/api/rest_v1/#/Feed/aggregatedFeed). We _could_ construct a timeline for all those forthcoming days, but the approach taken here instead is more DRY-style reuse of the existing data store/data the app already has rather than constructing a network stack in the widget. 

I wouldn't read too much into the `dateAtMidnight` stuff here for this widget, I'll probably end up ripping it out. Ideally, I think the refresh scenario for this widget will be something like this:
1. On main app's background refresh, reload this widget (work to do that already in here)
2. On any manual refresh of the explore feed in the main app, reload this widget
3. Every few hours (using the timeline refresh after policy mechanism above), refresh the widget.

I imagine a strategy like this might also be useful for the other widgets (this was why I made [WidgetController](https://github.com/wikimedia/wikipedia-ios/pull/3659/files#diff-40c9f930826459cea23eaa885cfa4f16) and defined the widget identifiers there – we can choose to reload only specific widgets here if we want to if some contextually relevant action occurs in the app).

Honestly the data modeling I'm using here is something I want to revisit again after I work on the Top read widget's data layer. I think there are probably some simplifications and changes I could make for the better after I dig into the Core Data store a little more. Feel free to share though if you figure out something that feels better suited for these widgets